lib/dry/container/mocking.rb

```diff
@@ -0,0 +1,78 @@
+module Dry
+  class Container
+    module Mocking
+      class Resolver < Container::Resolver
+        def initialize(resolver)
+          @resolver = resolver
+        end
+
+        def mocks
+          @mocks ||= {}
+        end
+
+        def call(original, key)
+          container = if mocks.key?(key.to_s)
+                        mocks
+                      else
+                        original
+                      end
+
+          @resolver.call(container, key)
+        end
+
+        def key?(container, key)
+          @resolver.key?(container, key)
+        end
+
+        def mock(registry, mocks, val = nil)
+          mocks = { mocks => val } unless mocks.is_a?(Hash)
+
+          mocks.each do |key, value|
+            registry.call(self.mocks, key, value, {})
+          end
+
+          mocks.keys
+        end
+
+        def unmock(keys)
+          keys = mocks.keys if keys.empty?
+
+          keys.map(&:to_s).each do |key|
+            mocks.delete(key)
+          end
+        end
+
+        def self.wrap(resolver)
+          case resolver
+          when self
+            resolver
+          else
+            new(resolver)
+          end
+        end
+      end
+    end
+
+    module Mixin
+      def mock(mocks, val = nil)
+        config.resolver = Mocking::Resolver.wrap(config.resolver)
```

Comment:
it can't be 100% consistent if we want to support mocking within a block as @solnic suggested.

What about something like:

```ruby
# spec_helper.rb
require 'dry/container/mocking'

# To inject mocking API into container
container.enable_mocking

# **/*_spec.rb
# Mock just like register
container.mock :key, :value
container.mock :key { something }

# to unmock a single key
container.unmock :key

# to "unmock" all mocked keys
container.unmock

# to mock within a block:
container.mocking key1: 'mock1', key2: 'mock2' do
  # Mocks for key1 and key2 only live here
  # Only key1 and key2 are cleared after
end
```